### PR TITLE
errors: introduce package errors module

### PR DESCRIPTION
**Pull request by cjp256**
- Introduce ProviderError, the error to encapsulate all craft-provider
  errors.

- Introduce details_from_called_process_error() to assist with handling
  subprocess.CalledProcessError exceptions.

- Introduce details_from_command_error() to assist with handling command
  errors that don't result in CalledProcessErrors (e.g. Popen()).

- Add unit tests for coverage.

- Remove test_todo.py as we now have a real test module.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: tests/unit/test_errors.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+
+def test_details_from_called_process_error():
+    error = subprocess.CalledProcessError(
+        -1, ["test-command", "flags", "quote$me"], "test stdout", "test stderr"
+    )
+
+    details = errors.details_from_called_process_error(error)
+
+    assert details == "\n".join(
```
Comment (sergiusens) [inline tests/unit/test_errors.py:60]: why not textwrap dedent instead?

Comment (cjp256) [reply tests/unit/test_errors.py:60]: I'll update to use it.

Comment (cjp256) [reply tests/unit/test_errors.py]: done.

Comment (cjp256) [reply tests/unit/test_errors.py]: (also added a decode() behavior for bytes with additional test case)

**Review by facundobatista — CHANGES_REQUESTED** [review]
Good, I like it! I annotated some details.

File: craft_providers/errors.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import subprocess
+from typing import List, Optional, Union
+
+
+def details_from_command_error(
+    *,
+    cmd: List[str],
+    returncode: int,
+    stdout: Optional[Union[bytes, str]] = None,
+    stderr: Optional[Union[bytes, str]] = None,
```
[the remark this answers is not in the document]
Comment (cjp256) [reply craft_providers/errors.py:27]: I tend to agree after adding the decode bits yesterday.  Output is not always a string, or decodable... This was an attempt to handle both cases consistently.

File: craft_providers/errors.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    If stdout or stderr have undecodable bytes, they will be updated using
+    decode(errors="replace").
+
+    :param cmd: Command executed.
+    :param stdout: Optional stdout to include.
+    :param stderr: Optional stderr to include.
+
+    :returns: Details string.
+    """
+    cmd_string = shlex.join(cmd)
```
[the remark this answers is not in the document]
Comment (cjp256) [reply craft_providers/errors.py:40]: People observe failed commands without the proper quoting and thinks the missing quotes is it's breaking (think parameters with spaces that now look like multiple args).  And I do copy commands occasionally :D

File: craft_providers/errors.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    details = [
+        f"* Command that failed: {cmd_string}",
```
[the remark this answers is not in the document]
Comment (cjp256) [reply craft_providers/errors.py:43]: As this output is for the user, I figured multi-line output would be the more user-friendly option.

File: craft_providers/errors.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjp256) [reply craft_providers/errors.py:27]: FWICT error.stdout will always be bytes unless requesting text mode (e.g. text=True).

File: craft_providers/errors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjp256) [reply craft_providers/errors.py]: Done

File: craft_providers/errors.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (facundobatista) [reply craft_providers/errors.py:27]: My point is that it should be responsibility of the generator of the info to do the decoding.

Having "bytes or unicode" traveling through the code is a bad idea. The conversion needs to be done ASAP, in the limit of where it's obtained (input) or print/sent (output).

So, IMO this function should only accept `str`. And it's responsibility of who's using `subprocess` to call `run`/`Popen`/whatever with `text=True` or do the decoding by hand.

[the remark this answers is not in the document]
Comment (cjp256) [reply craft_providers/errors.py:27]: I agree.  I dropped the decoding and now print the repr(), whether string or bytes.

**Review by facundobatista — APPROVED** [review]
Looks ok!

**Review by sergiusens — APPROVED** [review]
I am not sure how this came to be or how useful it might be, these methods feel more higher level than required for this library. Approving considering it is early days.

**Pull request merged**